compiler/rustc_lint/src/builtin.rs

```diff
@@ -2420,3 +2420,35 @@ impl<'tcx> LateLintPass<'tcx> for ClashingExternDeclarations {
         }
     }
 }
+
+declare_lint! {
+    FUNCTION_REFERENCES,
+    Warn,
+    "suggest casting functions to pointers when attempting to take references"
+}
+
+declare_lint_pass!(FunctionReferences => [FUNCTION_REFERENCES]);
+
+impl<'tcx> LateLintPass<'tcx> for FunctionReferences {
+    fn check_expr(&mut self, cx: &LateContext<'_>, e: &hir::Expr<'_>) {
+        if let hir::ExprKind::AddrOf(hir::BorrowKind::Ref, _, referent) = e.kind {
+            if let hir::ExprKind::Path(qpath) = &referent.kind {
+                if let Some(def_id) = cx.qpath_res(qpath, referent.hir_id).opt_def_id() {
+                    cx.tcx.hir().get_if_local(def_id).map(|node| {
```

Comment:
Hmm, why is this only looking in the current crate? I'd think we'd want to link on `&std::env::var` or something too.

Reply:
Good point, checking references to functions in other crates is probably important. I'm not sure I understand how to check other crates though. I've been looking through the docs for `hir::map::Map` and `ty::TyCtxt`, but if someone could give me a few pointers or help out that be great.

Reply:
I don't think this needs to work on the HIR at all. Try [`type_of`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.type_of) instead: if this is both a function and a path, that means you're taking a reference to it. `type_of` returns a [`Ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/type.Ty.html), so you could call `matches!(ty.kind, TyKind::FnDef(_, _))`: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.TyKind.html

Reply:
Yes, I think @jyn514 is right -- going through HIR here is unexpected.

Reply:
Actually, this is a design decision... do we want the lint to fire on
```rust
let fn_item = function_name;
let _val = &fn_item;
```
If yes, this should be type-based. If no, we have to instead check if the referent is the name of a function.